cloudfront_facts.py

wilvk opened the pull request:
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
cloud_front_facts

##### ANSIBLE VERSION
```
ansible 2.1.0.0
```

##### SUMMARY
This module allows for gathering of aws cloudfront facts using boto3. I initially developed it to get the dns name of a cloudfront distribution based on the CNAME alias. I decided to expand it to the full cloudfront facts once started. It was based on cloudformation_facts.py as a starting point. It's my first pull request so I hope it's ok. :)

It gathers facts on individual elements (get-) as well as lists, such as:

get-cloud-front-origin-access-identity
get-cloud-front-origin-access-identity-config
get-distribution
get-distribution-config
get-invalidation
get-streaming-distribution
get-streaming-distribution-config
list-cloud-front-origin-access-identities
list-distributions
list-distributions-by-web-acl-id
list-invalidations
list-streaming-distributions

It uses the following boolean inputs to specify the output:

- distribution
- distribution_config
- cloud_front_origin_access_identity
- cloud_front_origin_access_identity_config
- invalidation
- get_streaming_distribution
- get_streaming_distribution  

And the following string inputs to specify details of the facts:

- distribution_id or the domain_name_alias 
- invalidation_id 
- cloud_front_origin_access_identity_id
- web_acl_id

Commit:
initial commit of cloudfront_facts.py

georgepsarakis commented on lib/ansible/modules/cloud/amazon/cloudfront_facts.py:
```diff
@@ -0,0 +1,453 @@
+def to_dict(items, key, value):
+    ''' Transforms a list of items to a Key/Value dictionary '''
+    if items:
+        return dict(zip([i[key] for i in items], [i[value] for i in items]))
```
If you allow me a suggestion here:

```python
def _extract_key_from_object_list(items, key, value):
    return dict((item[key], item[value]) for item in items)
```

You do not need to add the if-else construct, unless of course `items` may not be iterable.

Also, I don't see any usage of this helper function in this module, perhaps it is not needed after all?

georgepsarakis commented on lib/ansible/modules/cloud/amazon/cloudfront_facts.py:
```diff
@@ -0,0 +1,453 @@
+        facts['list_invalidations'] = service_mgr.list_invalidations(distribution_id)
+
+    result['changed'] = False
+    module.exit_json(msg="Retreived cloudfront facts.", ansible_facts=result)
```
Small typo here probably: `Retreived -> Retrieved`

wilvk replied:
Thanks for the review @georgepsarakis! I'll remove that function as it is not used. Have also fixed the typo, and will do a pull request to https://github.com/ansible/ansible-modules-core/pulls

wilvk replied:
Tried pull request to ansible-modules-core and got the following message:

This repository is locked

Please open all new issues and pull requests in https://github.com/ansible/ansible

For more information please see http://docs.ansible.com/ansible/dev_guide/repomerge.html

georgepsarakis commented:
@wilvk I think pull requests regarding core Ansible modules should be opened in this repository: https://github.com/ansible/ansible-modules-core/pulls

Commit:
modification as per review from @georgepsarakis

georgepsarakis commented:
@wilvk sorry did not know it myself, it is quite recent! 

> On Tuesday 6th December 2016, the Ansible Core Team re-merged the module repositories back into ansible/ansible in GitHub.

wilvk committed:
fixed shippable build error

wilvk committed:
fixed shippable build error

georgepsarakis commented on lib/ansible/modules/cloud/amazon/cloudfront_facts.py:
```diff
@@ -0,0 +1,448 @@
+
+# import module snippets
+from ansible.module_utils.basic import *
+from ansible.module_utils.ec2 import *
```
I think that wildcard imports are generally discouraged, because it makes it difficult to refactor and determine the origin of a function or class. For example, at first the `boto3_conn` function used in the constructor seemed undefined.

wilvk replied:
Have removed the wildcard import and only imported the 2 classes required.

georgepsarakis commented on lib/ansible/modules/cloud/amazon/cloudfront_facts.py:
```diff
@@ -0,0 +1,448 @@
+        facts['distribution'] = service_mgr.get_distribution(distribution_id)
+    if distribution_config:
+        facts['distribution_config'] = service_mgr.get_distribution_config(distribution_id)
+    if cloud_front_origin_access_identity:
```
Perhaps you could drop the `cloud_front_` prefix for methods and variables, since everything refers to Cloudfront properties? Unless of course it is has something to do with being consistent with the corresponding AWS API response.

wilvk replied:
I agree. The `cloud_front_` prefix was to conform with the[ boto3 api ](http://boto3.readthedocs.io/en/latest/reference/services/cloudfront.html#CloudFront.Client.get_cloud_front_origin_access_identity) but it is superfluous in the context of a cloud front module. Even the AWS GUI calls it 'Origin Access Identity' without the prefix. Have renamed all instances.

Commit:
removed wildcard imports and replaced with specific imports from ansible.module_utils.ec2 as advised by @georgepsarakis

Commit:
renamed all instances of cloud_front_origin_access_identity to origin_access_identity as advised by @georgepsarakis

Commit:
added input option requirements where missing. fixed a typo in an error message, added '.' to the end of sentences.

Commit:
fixed typo in documentation

georgepsarakis commented on lib/ansible/modules/cloud/amazon/cloudfront_facts.py:
```diff
@@ -0,0 +1,446 @@
+        except Exception as e:
+            self.module.fail_json(msg="Error getting distribution id from domain name = " + str(e), exception=traceback.format_exc(e))
+
+    def paginated_response(self, func, result_key, next_token=None):
```
Glad if I helped so far. By the way, this is a very helpful module! If you allow me one final note, I would suggest to avoid recursion since it is a difficult subject in general. You could achieve the same effect here, using a `while` loop:

```python
def paginated_response(func, result_key):
    kwargs = dict()
    results = []
    loop = True
    while loop:
        response = func(**kwargs)
        result = response.get(result_key)
        results.extend(result)
        kwargs['NextToken'] = response.get('NextToken')
        loop = kwargs['NextToken'] is not None
    return results    
```

What do you think?

wilvk replied:
Thanks, @georgepsarakis, your feedback is very welcome! Yes, that is a good idea. Will implement it.

ryansb commented on lib/ansible/modules/cloud/amazon/cloudfront_facts.py:
```diff
@@ -0,0 +1,446 @@
+        facts['list_invalidations'] = service_mgr.list_invalidations(distribution_id)
+
+    result['changed'] = False
+    module.exit_json(msg="Retrieved cloudfront facts.", ansible_facts=result)
```
It looks like you're double-nesting `ansible_facts` dictionary, on line 406/409/412 you have the facts being saved in `result = { 'ansible_facts': { 'cloudfront':...` but then here you pass to exit_json under `ansible_facts` so your output looks like:

```
localhost | SUCCESS => {                                                                                                          [479/1550]
    "ansible_facts": {
        "ansible_facts": {
            "cloudfront": {
                "E22CS9R7XQ0CWU": {
                    "distribution": {
....
```
To fix this you can instead do `module.exit_json(msg="Retrieved cloudfront facts.", **result)` or stop nesting the `cloudfront` key under `ansible_facts` on 402/405/408/411.

wilvk replied:
Thanks for the feedback @ryansb! All good suggestions. I'll set the default to list_distributions, fix the double-nesting and version number and list both distribution_id and CNAME where possible.

wilvk replied:
Thanks again for the feedback - I've implemented all the suggestions. 

For listing by both distribution_id and alias, if the distribution_id had been set (not necessarily as a parameter), it queries list_distributions for the aliases and lists by these as well. If only domain_name_alias is specified, this make it 3 calls on the boto3 api (1 to get distribution_id, 1 to get aliases and 1 for the actual request) making it a bit 'chatty' but still a reasonable response time imho. The alternative was to use only the domain_name_alias parameter but this may have missed other aliases of the distribution.

Also, the alias listing has only been done for distribution and distribution_config but distribution_id is used for other options as well (eg. invalidation, streaming_distribution and streaming_distribution_config) and so the aliases for these could also be used if required.

wilvk replied:
Hi, I found a few bugs in my code which I have since fixed:
- default to list_distributions logic was incorrect - check on if list_distributions is specified
- list_distributions_by_web_acl wasn't passing web_acl_id and wasn't specifying WebAclId for boto3 kwargs
- list_invalidations - wasn't specifying DistributionId for kwargs to boto3

ryansb commented on lib/ansible/modules/cloud/amazon/cloudfront_facts.py:
```diff
@@ -0,0 +1,446 @@
+    # set appropriate ansible_facts id
+    if distribution_id:
+        result = { 'ansible_facts': { 'cloudfront': { distribution_id:{} } } }
+        facts = result['ansible_facts']['cloudfront'][distribution_id]
```
Nesting under distribution ID is useful when you know it, but when returning a single item it's nice if you can make it available under the key the user asked for it with. Take for example: `ansible -m cloudfront_facts -a 'region=us-east-1 distribution=yes domain_name_alias=rsb.io' localhost`. As a user, I'd like to be able to say something like: `{{ cloudfront['rsb.io].ARN }}` to get the ARN, instead of (currently) `{{ cloudfront.[cloudfront.keys()[0]].distribution.ARN }}`. You could return both, so users could either use the ID or one of the CNAMEs to access the facts. They would be unique, since in CloudFront you can't have overlapping CNAMEs ever.

ryansb commented on lib/ansible/modules/cloud/amazon/cloudfront_facts.py:
```diff
@@ -0,0 +1,446 @@
+requirements:
+  - boto3 >= 1.0.0
+  - python >= 2.6
+version_added: "2.2"
```
This will need to be bumpet up to `2.3`

wilvk replied:
Good idea. Will add it.

ryansb reviewed (changes requested):
Thank you for writing this module, it'd be great to have it in 2.3. Is there a way to make the default behavior a little more useful to new folks? As a new user of this module I tried:

```
ansible -m cloudfront_facts -a 'profile=personal region=us-east-1' localhost
 [WARNING]: Host file not found: /etc/ansible/hosts

 [WARNING]: provided hosts list is empty, only localhost is available

localhost | SUCCESS => {
    "ansible_facts": {
        "ansible_facts": {
            "cloudfront": {}
        }, 
        "changed": false
    }, 
    "changed": false, 
    "msg": "Retrieved cloudfront facts."
}
```
But there was no output, maybe if nothing is provided defaulting to `list_distributions=yes` would be easiest, so users can see some facts with just the default usage.

Commit:
simplified paginated_response method as advised by @georgepsarakis

Commit:
set default option to list_defaults when no option specified as advised by @ryansb

Commit:
bumped version to 2.3 as advised by @ryansb

Commit:
removed double-nesting of ansible_facts as advised by @ryansb

Commit:
show facts based on alias and distribution id for easy referencing as advised by @ryansb. have done for both distribution and distribution_config

Commit:
made comments clearer

Commit:
fixed incorrect logic for default list_distributions, fixed list_distributions_by_web_acl - wasn't passing web_acl_id, fixed list_invalidations keyword args missing DistributionId

Commit:
neatened up logic for list_distributions default

Commit:
facts now return Etag as well as specific facts. modified paginated_response to handle this. also all requests that use distribution id also list the facts under the corresponding alias now.

wilvk commented:
I had to make a few changes to the module - I hope you don't mind. 
- Each distribution has an ETag that identifies the latest revision of the distribution. This wasn't being returned as the paginated_response method was getting a specific key from the response for the get_ queries and the responses with an ETag have multiple top-level keys (eg. 'Distribution' and 'Etag'). For these get_ queries, the name was part of the response (eg Distribution) and so it is double-nesting the name under the relevant section as this was being explicitly defined (similar to the ansible_facts nesting issue highlighted earlier).
- For invalidation, streaming_distribution and streaming_distribution_config, I have added facts for the aliased domain names, and for invalidation I have also added invalidation_id for completeness.

Commit:
fixed last fact added clobbering all previous facts for an alias or an id

ryansb commented:
It looks like the defaults (using `ansible -m cloudfront_facts -a 'profile=personal region=us-east-1' localhost`) returns a list, but the list is just the straight boto3 response, so it's still nested and doesn't let you do things like `{{ cloudfront.distributions | keys() }}` to get all the names/IDs available.

wilvk commented:
Yep, no worries - will do. 
Just to confirm, I'll move everything under `ansible_facts::cloudfront::list_distributions::Items[]` to be directly under `ansible_facts::cloudfront::list_distributions` as per below. I'll do this for all the `list_` calls as they all appear to be affected.

```
{
    "ansible_facts": {
        "changed": false,
        "cloudfront": {
            "list_distributions": {
                "IsTruncated": false,
                "Items": [
                    {
                        "ARN": "arn:aws:cloudfront::111111111111:distribution/E3ABCDEFGJIJKL",
                        "Aliases": {
                            "Items": [
                                "rsb.io"
                            ],
                            "Quantity": 1

```

I was also thinking of adding a summary of all the distribution ids, invalidation ids, origin access ids and aliases in a format similar to the following that could be used as the default instead of `list_distributions`, or a `summary:yes` option which would be useful for new users:

```
{ 
    "distributions":  [  
        "distribution_id": "E1ABCDEFGHILKL",
        "aliases": [ 
            "rsb.io", 
            "seso.io",
            ...
        ],  
        "invalidations": [
            "IN9ABCDEFGHIJ",
            "IN9ABCDEABCDE",
            ...
        ]
    ],
    "origin_access_identity_ids": [
        "E2ABCDEFGHIJKL",
        "E2ABCDEFABCDEF",
        ...
    ]
}
```
Let me know what you think.

Commit:
removed list_ prefix from list keys

Commit:
removed unnecessary boto fields. made list_distributions and list_streaming_distributions dictionaries with id/alias as key. fixed list_invalidations.

Commit:
fixed incorrectly named list_distributions_by_web_acl to ..web_acl_id. added id/alias dict for list_distributions_by_web_acl_id

Commit:
minor doc changes

Commit:
fixed merge issue

Commit:
removed merge tag

Commit:
fixed more of merge

ryansb commented:
Looks great, thank you for contributing!

No worries about the summary info, that can come in a follow-on PR if you have time to flesh it out later.

ryansb commented:
Here's a real output example, btw:

```
{
  "ansible_facts": {
    "changed": false,
    "cloudfront": {
      "distributions": {
        "*.serverlesscode.com": {
          "ARN": "arn:aws:cloudfront::123456789012:distribution/E22CS9R7XQ0CWU",
          "Aliases": {
            "Items": [
              "serverlesscode.com",
              "*.serverlesscode.com"
            ],
            "Quantity": 2
          },
          "CacheBehaviors": {
            "Quantity": 0
          },
          "Comment": "",
          "CustomErrorResponses": {
            "Items": [
              {
                "ErrorCachingMinTTL": 10,
                "ErrorCode": 400,
                "ResponseCode": "",
                "ResponsePagePath": ""
              },
              {
                "ErrorCachingMinTTL": 30,
                "ErrorCode": 403,
                "ResponseCode": "",
                "ResponsePagePath": ""
              },
              {
                "ErrorCachingMinTTL": 30,
                "ErrorCode": 404,
                "ResponseCode": "",
                "ResponsePagePath": ""
              }
            ],
            "Quantity": 3
          },
          "DefaultCacheBehavior": {
            "AllowedMethods": {
              "CachedMethods": {
                "Items": [
                  "HEAD",
                  "GET"
                ],
                "Quantity": 2
              },
              "Items": [
                "HEAD",
                "GET"
              ],
              "Quantity": 2
            },
            "Compress": true,
            "DefaultTTL": 1800,
            "ForwardedValues": {
              "Cookies": {
                "Forward": "none"
              },
              "Headers": {
                "Quantity": 0
              },
              "QueryString": true,
              "QueryStringCacheKeys": {
                "Items": [
                  "h"
                ],
                "Quantity": 1
              }
            },
            "LambdaFunctionAssociations": {
              "Quantity": 0
            },
            "MaxTTL": 31536000,
            "MinTTL": 180,
            "SmoothStreaming": false,
            "TargetOriginId": "s3origin",
            "TrustedSigners": {
              "Enabled": false,
              "Quantity": 0
            },
            "ViewerProtocolPolicy": "redirect-to-https"
          },
          "DomainName": "d1nn5majnuzzvm.cloudfront.net",
          "Enabled": true,
          "Id": "E22CS9R7XQ0CWU",
          "LastModifiedTime": "2016-12-19T21:30:05.438000+00:00",
          "Origins": {
            "Items": [
              {
                "CustomHeaders": {
                  "Quantity": 0
                },
                "CustomOriginConfig": {
                  "HTTPPort": 80,
                  "HTTPSPort": 443,
                  "OriginProtocolPolicy": "http-only",
                  "OriginSslProtocols": {
                    "Items": [
                      "SSLv3",
                      "TLSv1"
                    ],
                    "Quantity": 2
                  }
                },
                "DomainName": "production.serverlesscode.com.s3-website-us-east-1.amazonaws.com",
                "Id": "s3origin",
                "OriginPath": ""
              }
            ],
            "Quantity": 1
          },
          "PriceClass": "PriceClass_200",
          "Restrictions": {
            "GeoRestriction": {
              "Quantity": 0,
              "RestrictionType": "none"
            }
          },
          "Status": "Deployed",
          "ViewerCertificate": {
            "ACMCertificateArn": "arn:aws:acm:us-east-1:123456789012:certificate/9e6733ba-cad8-4624-b58f-416c93685e2c",
            "Certificate": "arn:aws:acm:us-east-1:123456789012:certificate/9e6733ba-cad8-4624-b58f-416c93685e2c",
            "CertificateSource": "acm",
            "MinimumProtocolVersion": "TLSv1",
            "SSLSupportMethod": "sni-only"
          },
          "WebACLId": ""
        },
        "E22CS9R7XQ0CWU": {
          "ARN": "arn:aws:cloudfront::123456789012:distribution/E22CS9R7XQ0CWU",
          "Aliases": {
            "Items": [
              "serverlesscode.com",
              "*.serverlesscode.com"
            ],
            "Quantity": 2
          },
          "CacheBehaviors": {
            "Quantity": 0
          },
          "Comment": "",
          "CustomErrorResponses": {
            "Items": [
              {
                "ErrorCachingMinTTL": 10,
                "ErrorCode": 400,
                "ResponseCode": "",
                "ResponsePagePath": ""
              },
              {
                "ErrorCachingMinTTL": 30,
                "ErrorCode": 403,
                "ResponseCode": "",
                "ResponsePagePath": ""
              },
              {
                "ErrorCachingMinTTL": 30,
                "ErrorCode": 404,
                "ResponseCode": "",
                "ResponsePagePath": ""
              }
            ],
            "Quantity": 3
          },
          "DefaultCacheBehavior": {
            "AllowedMethods": {
              "CachedMethods": {
                "Items": [
                  "HEAD",
                  "GET"
                ],
                "Quantity": 2
              },
              "Items": [
                "HEAD",
                "GET"
              ],
              "Quantity": 2
            },
            "Compress": true,
            "DefaultTTL": 1800,
            "ForwardedValues": {
              "Cookies": {
                "Forward": "none"
              },
              "Headers": {
                "Quantity": 0
              },
              "QueryString": true,
              "QueryStringCacheKeys": {
                "Items": [
                  "h"
                ],
                "Quantity": 1
              }
            },
            "LambdaFunctionAssociations": {
              "Quantity": 0
            },
            "MaxTTL": 31536000,
            "MinTTL": 180,
            "SmoothStreaming": false,
            "TargetOriginId": "s3origin",
            "TrustedSigners": {
              "Enabled": false,
              "Quantity": 0
            },
            "ViewerProtocolPolicy": "redirect-to-https"
          },
          "DomainName": "d1nn5majnuzzvm.cloudfront.net",
          "Enabled": true,
          "Id": "E22CS9R7XQ0CWU",
          "LastModifiedTime": "2016-12-19T21:30:05.438000+00:00",
          "Origins": {
            "Items": [
              {
                "CustomHeaders": {
                  "Quantity": 0
                },
                "CustomOriginConfig": {
                  "HTTPPort": 80,
                  "HTTPSPort": 443,
                  "OriginProtocolPolicy": "http-only",
                  "OriginSslProtocols": {
                    "Items": [
                      "SSLv3",
                      "TLSv1"
                    ],
                    "Quantity": 2
                  }
                },
                "DomainName": "production.serverlesscode.com.s3-website-us-east-1.amazonaws.com",
                "Id": "s3origin",
                "OriginPath": ""
              }
            ],
            "Quantity": 1
          },
          "PriceClass": "PriceClass_200",
          "Restrictions": {
            "GeoRestriction": {
              "Quantity": 0,
              "RestrictionType": "none"
            }
          },
          "Status": "Deployed",
          "ViewerCertificate": {
            "ACMCertificateArn": "arn:aws:acm:us-east-1:123456789012:certificate/9e6733ba-cad8-4624-b58f-416c93685e2c",
            "Certificate": "arn:aws:acm:us-east-1:123456789012:certificate/9e6733ba-cad8-4624-b58f-416c93685e2c",
            "CertificateSource": "acm",
            "MinimumProtocolVersion": "TLSv1",
            "SSLSupportMethod": "sni-only"
          },
          "WebACLId": ""
        },
        "E26RZ9XRFCAYV1": {
          "ARN": "arn:aws:cloudfront::123456789012:distribution/E26RZ9XRFCAYV1",
          "Aliases": {
            "Items": [
              "decau.se"
            ],
            "Quantity": 1
          },
          "CacheBehaviors": {
            "Quantity": 0
          },
          "Comment": "",
          "CustomErrorResponses": {
            "Items": [
              {
                "ErrorCachingMinTTL": 10,
                "ErrorCode": 400,
                "ResponseCode": "",
                "ResponsePagePath": ""
              },
              {
                "ErrorCachingMinTTL": 30,
                "ErrorCode": 403,
                "ResponseCode": "",
                "ResponsePagePath": ""
              },
              {
                "ErrorCachingMinTTL": 30,
                "ErrorCode": 404,
                "ResponseCode": "",
                "ResponsePagePath": ""
              }
            ],
            "Quantity": 3
          },
          "DefaultCacheBehavior": {
            "AllowedMethods": {
              "CachedMethods": {
                "Items": [
                  "HEAD",
                  "GET"
                ],
                "Quantity": 2
              },
              "Items": [
                "HEAD",
                "GET"
              ],
              "Quantity": 2
            },
            "Compress": true,
            "DefaultTTL": 10800,
            "ForwardedValues": {
              "Cookies": {
                "Forward": "none"
              },
              "Headers": {
                "Quantity": 0
              },
              "QueryString": true,
              "QueryStringCacheKeys": {
                "Quantity": 0
              }
            },
            "LambdaFunctionAssociations": {
              "Quantity": 0
            },
            "MaxTTL": 2592000,
            "MinTTL": 90,
            "SmoothStreaming": false,
            "TargetOriginId": "s3origin",
            "TrustedSigners": {
              "Enabled": false,
              "Quantity": 0
            },
            "ViewerProtocolPolicy": "redirect-to-https"
          },
          "DomainName": "d3cdqd9adwvuex.cloudfront.net",
          "Enabled": true,
          "Id": "E26RZ9XRFCAYV1",
          "LastModifiedTime": "2016-10-14T23:03:26.268000+00:00",
          "Origins": {
            "Items": [
              {
                "CustomHeaders": {
                  "Quantity": 0
                },
                "CustomOriginConfig": {
                  "HTTPPort": 80,
                  "HTTPSPort": 443,
                  "OriginProtocolPolicy": "http-only",
                  "OriginSslProtocols": {
                    "Items": [
                      "SSLv3",
                      "TLSv1"
                    ],
                    "Quantity": 2
                  }
                },
                "DomainName": "decause-staticbucket-1pbtkxd1ua1zh.s3-website-us-east-1.amazonaws.com",
                "Id": "s3origin",
                "OriginPath": ""
              }
            ],
            "Quantity": 1
          },
          "PriceClass": "PriceClass_All",
          "Restrictions": {
            "GeoRestriction": {
              "Quantity": 0,
              "RestrictionType": "none"
            }
          },
          "Status": "Deployed",
          "ViewerCertificate": {
            "ACMCertificateArn": "arn:aws:acm:us-east-1:123456789012:certificate/cef38833-c8d2-4c81-aeb6-4f833037c8dd",
            "Certificate": "arn:aws:acm:us-east-1:123456789012:certificate/cef38833-c8d2-4c81-aeb6-4f833037c8dd",
            "CertificateSource": "acm",
            "MinimumProtocolVersion": "TLSv1",
            "SSLSupportMethod": "sni-only"
          },
          "WebACLId": ""
        },
        "E3EUUQMGDFJ3JH": {
          "ARN": "arn:aws:cloudfront::123456789012:distribution/E3EUUQMGDFJ3JH",
          "Aliases": {
            "Items": [
              "rsb.io"
            ],
            "Quantity": 1
          },
          "CacheBehaviors": {
            "Quantity": 0
          },
          "Comment": "",
          "CustomErrorResponses": {
            "Items": [
              {
                "ErrorCachingMinTTL": 10,
                "ErrorCode": 400,
                "ResponseCode": "",
                "ResponsePagePath": ""
              },
              {
                "ErrorCachingMinTTL": 30,
                "ErrorCode": 403,
                "ResponseCode": "",
                "ResponsePagePath": ""
              },
              {
                "ErrorCachingMinTTL": 30,
                "ErrorCode": 404,
                "ResponseCode": "",
                "ResponsePagePath": ""
              }
            ],
            "Quantity": 3
          },
          "DefaultCacheBehavior": {
            "AllowedMethods": {
              "CachedMethods": {
                "Items": [
                  "HEAD",
                  "GET"
                ],
                "Quantity": 2
              },
              "Items": [
                "HEAD",
                "GET"
              ],
              "Quantity": 2
            },
            "Compress": true,
            "DefaultTTL": 3600,
            "ForwardedValues": {
              "Cookies": {
                "Forward": "none"
              },
              "Headers": {
                "Quantity": 0
              },
              "QueryString": true,
              "QueryStringCacheKeys": {
                "Quantity": 0
              }
            },
            "LambdaFunctionAssociations": {
              "Quantity": 0
            },
            "MaxTTL": 2592000,
            "MinTTL": 60,
            "SmoothStreaming": false,
            "TargetOriginId": "s3origin",
            "TrustedSigners": {
              "Enabled": false,
              "Quantity": 0
            },
            "ViewerProtocolPolicy": "redirect-to-https"
          },
          "DomainName": "d2o396wjt6ehzs.cloudfront.net",
          "Enabled": true,
          "Id": "E3EUUQMGDFJ3JH",
          "LastModifiedTime": "2016-12-19T21:30:32.281000+00:00",
          "Origins": {
            "Items": [
              {
                "CustomHeaders": {
                  "Quantity": 0
                },
                "CustomOriginConfig": {
                  "HTTPPort": 80,
                  "HTTPSPort": 443,
                  "OriginProtocolPolicy": "http-only",
                  "OriginSslProtocols": {
                    "Items": [
                      "SSLv3",
                      "TLSv1"
                    ],
                    "Quantity": 2
                  }
                },
                "DomainName": "rsbio-staticbucket-18zabyhgogmp.s3-website-us-east-1.amazonaws.com",
                "Id": "s3origin",
                "OriginPath": ""
              }
            ],
            "Quantity": 1
          },
          "PriceClass": "PriceClass_All",
          "Restrictions": {
            "GeoRestriction": {
              "Quantity": 0,
              "RestrictionType": "none"
            }
          },
          "Status": "Deployed",
          "ViewerCertificate": {
            "ACMCertificateArn": "arn:aws:acm:us-east-1:123456789012:certificate/bc4f1eca-c800-4458-8db1-8d76afdc6f44",
            "Certificate": "arn:aws:acm:us-east-1:123456789012:certificate/bc4f1eca-c800-4458-8db1-8d76afdc6f44",
            "CertificateSource": "acm",
            "MinimumProtocolVersion": "TLSv1",
            "SSLSupportMethod": "sni-only"
          },
          "WebACLId": ""
        },
        "decau.se": {
          "ARN": "arn:aws:cloudfront::123456789012:distribution/E26RZ9XRFCAYV1",
          "Aliases": {
            "Items": [
              "decau.se"
            ],
            "Quantity": 1
          },
          "CacheBehaviors": {
            "Quantity": 0
          },
          "Comment": "",
          "CustomErrorResponses": {
            "Items": [
              {
                "ErrorCachingMinTTL": 10,
                "ErrorCode": 400,
                "ResponseCode": "",
                "ResponsePagePath": ""
              },
              {
                "ErrorCachingMinTTL": 30,
                "ErrorCode": 403,
                "ResponseCode": "",
                "ResponsePagePath": ""
              },
              {
                "ErrorCachingMinTTL": 30,
                "ErrorCode": 404,
                "ResponseCode": "",
                "ResponsePagePath": ""
              }
            ],
            "Quantity": 3
          },
          "DefaultCacheBehavior": {
            "AllowedMethods": {
              "CachedMethods": {
                "Items": [
                  "HEAD",
                  "GET"
                ],
                "Quantity": 2
              },
              "Items": [
                "HEAD",
                "GET"
              ],
              "Quantity": 2
            },
            "Compress": true,
            "DefaultTTL": 10800,
            "ForwardedValues": {
              "Cookies": {
                "Forward": "none"
              },
              "Headers": {
                "Quantity": 0
              },
              "QueryString": true,
              "QueryStringCacheKeys": {
                "Quantity": 0
              }
            },
            "LambdaFunctionAssociations": {
              "Quantity": 0
            },
            "MaxTTL": 2592000,
            "MinTTL": 90,
            "SmoothStreaming": false,
            "TargetOriginId": "s3origin",
            "TrustedSigners": {
              "Enabled": false,
              "Quantity": 0
            },
            "ViewerProtocolPolicy": "redirect-to-https"
          },
          "DomainName": "d3cdqd9adwvuex.cloudfront.net",
          "Enabled": true,
          "Id": "E26RZ9XRFCAYV1",
          "LastModifiedTime": "2016-10-14T23:03:26.268000+00:00",
          "Origins": {
            "Items": [
              {
                "CustomHeaders": {
                  "Quantity": 0
                },
                "CustomOriginConfig": {
                  "HTTPPort": 80,
                  "HTTPSPort": 443,
                  "OriginProtocolPolicy": "http-only",
                  "OriginSslProtocols": {
                    "Items": [
                      "SSLv3",
                      "TLSv1"
                    ],
                    "Quantity": 2
                  }
                },
                "DomainName": "decause-staticbucket-1pbtkxd1ua1zh.s3-website-us-east-1.amazonaws.com",
                "Id": "s3origin",
                "OriginPath": ""
              }
            ],
            "Quantity": 1
          },
          "PriceClass": "PriceClass_All",
          "Restrictions": {
            "GeoRestriction": {
              "Quantity": 0,
              "RestrictionType": "none"
            }
          },
          "Status": "Deployed",
          "ViewerCertificate": {
            "ACMCertificateArn": "arn:aws:acm:us-east-1:123456789012:certificate/cef38833-c8d2-4c81-aeb6-4f833037c8dd",
            "Certificate": "arn:aws:acm:us-east-1:123456789012:certificate/cef38833-c8d2-4c81-aeb6-4f833037c8dd",
            "CertificateSource": "acm",
            "MinimumProtocolVersion": "TLSv1",
            "SSLSupportMethod": "sni-only"
          },
          "WebACLId": ""
        },
        "rsb.io": {
          "ARN": "arn:aws:cloudfront::123456789012:distribution/E3EUUQMGDFJ3JH",
          "Aliases": {
            "Items": [
              "rsb.io"
            ],
            "Quantity": 1
          },
          "CacheBehaviors": {
            "Quantity": 0
          },
          "Comment": "",
          "CustomErrorResponses": {
            "Items": [
              {
                "ErrorCachingMinTTL": 10,
                "ErrorCode": 400,
                "ResponseCode": "",
                "ResponsePagePath": ""
              },
              {
                "ErrorCachingMinTTL": 30,
                "ErrorCode": 403,
                "ResponseCode": "",
                "ResponsePagePath": ""
              },
              {
                "ErrorCachingMinTTL": 30,
                "ErrorCode": 404,
                "ResponseCode": "",
                "ResponsePagePath": ""
              }
            ],
            "Quantity": 3
          },
          "DefaultCacheBehavior": {
            "AllowedMethods": {
              "CachedMethods": {
                "Items": [
                  "HEAD",
                  "GET"
                ],
                "Quantity": 2
              },
              "Items": [
                "HEAD",
                "GET"
              ],
              "Quantity": 2
            },
            "Compress": true,
            "DefaultTTL": 3600,
            "ForwardedValues": {
              "Cookies": {
                "Forward": "none"
              },
              "Headers": {
                "Quantity": 0
              },
              "QueryString": true,
              "QueryStringCacheKeys": {
                "Quantity": 0
              }
            },
            "LambdaFunctionAssociations": {
              "Quantity": 0
            },
            "MaxTTL": 2592000,
            "MinTTL": 60,
            "SmoothStreaming": false,
            "TargetOriginId": "s3origin",
            "TrustedSigners": {
              "Enabled": false,
              "Quantity": 0
            },
            "ViewerProtocolPolicy": "redirect-to-https"
          },
          "DomainName": "d2o396wjt6ehzs.cloudfront.net",
          "Enabled": true,
          "Id": "E3EUUQMGDFJ3JH",
          "LastModifiedTime": "2016-12-19T21:30:32.281000+00:00",
          "Origins": {
            "Items": [
              {
                "CustomHeaders": {
                  "Quantity": 0
                },
                "CustomOriginConfig": {
                  "HTTPPort": 80,
                  "HTTPSPort": 443,
                  "OriginProtocolPolicy": "http-only",
                  "OriginSslProtocols": {
                    "Items": [
                      "SSLv3",
                      "TLSv1"
                    ],
                    "Quantity": 2
                  }
                },
                "DomainName": "rsbio-staticbucket-18zabyhgogmp.s3-website-us-east-1.amazonaws.com",
                "Id": "s3origin",
                "OriginPath": ""
              }
            ],
            "Quantity": 1
          },
          "PriceClass": "PriceClass_All",
          "Restrictions": {
            "GeoRestriction": {
              "Quantity": 0,
              "RestrictionType": "none"
            }
          },
          "Status": "Deployed",
          "ViewerCertificate": {
            "ACMCertificateArn": "arn:aws:acm:us-east-1:123456789012:certificate/bc4f1eca-c800-4458-8db1-8d76afdc6f44",
            "Certificate": "arn:aws:acm:us-east-1:123456789012:certificate/bc4f1eca-c800-4458-8db1-8d76afdc6f44",
            "CertificateSource": "acm",
            "MinimumProtocolVersion": "TLSv1",
            "SSLSupportMethod": "sni-only"
          },
          "WebACLId": ""
        },
        "serverlesscode.com": {
          "ARN": "arn:aws:cloudfront::123456789012:distribution/E22CS9R7XQ0CWU",
          "Aliases": {
            "Items": [
              "serverlesscode.com",
              "*.serverlesscode.com"
            ],
            "Quantity": 2
          },
          "CacheBehaviors": {
            "Quantity": 0
          },
          "Comment": "",
          "CustomErrorResponses": {
            "Items": [
              {
                "ErrorCachingMinTTL": 10,
                "ErrorCode": 400,
                "ResponseCode": "",
                "ResponsePagePath": ""
              },
              {
                "ErrorCachingMinTTL": 30,
                "ErrorCode": 403,
                "ResponseCode": "",
                "ResponsePagePath": ""
              },
              {
                "ErrorCachingMinTTL": 30,
                "ErrorCode": 404,
                "ResponseCode": "",
                "ResponsePagePath": ""
              }
            ],
            "Quantity": 3
          },
          "DefaultCacheBehavior": {
            "AllowedMethods": {
              "CachedMethods": {
                "Items": [
                  "HEAD",
                  "GET"
                ],
                "Quantity": 2
              },
              "Items": [
                "HEAD",
                "GET"
              ],
              "Quantity": 2
            },
            "Compress": true,
            "DefaultTTL": 1800,
            "ForwardedValues": {
              "Cookies": {
                "Forward": "none"
              },
              "Headers": {
                "Quantity": 0
              },
              "QueryString": true,
              "QueryStringCacheKeys": {
                "Items": [
                  "h"
                ],
                "Quantity": 1
              }
            },
            "LambdaFunctionAssociations": {
              "Quantity": 0
            },
            "MaxTTL": 31536000,
            "MinTTL": 180,
            "SmoothStreaming": false,
            "TargetOriginId": "s3origin",
            "TrustedSigners": {
              "Enabled": false,
              "Quantity": 0
            },
            "ViewerProtocolPolicy": "redirect-to-https"
          },
          "DomainName": "d1nn5majnuzzvm.cloudfront.net",
          "Enabled": true,
          "Id": "E22CS9R7XQ0CWU",
          "LastModifiedTime": "2016-12-19T21:30:05.438000+00:00",
          "Origins": {
            "Items": [
              {
                "CustomHeaders": {
                  "Quantity": 0
                },
                "CustomOriginConfig": {
                  "HTTPPort": 80,
                  "HTTPSPort": 443,
                  "OriginProtocolPolicy": "http-only",
                  "OriginSslProtocols": {
                    "Items": [
                      "SSLv3",
                      "TLSv1"
                    ],
                    "Quantity": 2
                  }
                },
                "DomainName": "production.serverlesscode.com.s3-website-us-east-1.amazonaws.com",
                "Id": "s3origin",
                "OriginPath": ""
              }
            ],
            "Quantity": 1
          },
          "PriceClass": "PriceClass_200",
          "Restrictions": {
            "GeoRestriction": {
              "Quantity": 0,
              "RestrictionType": "none"
            }
          },
          "Status": "Deployed",
          "ViewerCertificate": {
            "ACMCertificateArn": "arn:aws:acm:us-east-1:123456789012:certificate/9e6733ba-cad8-4624-b58f-416c93685e2c",
            "Certificate": "arn:aws:acm:us-east-1:123456789012:certificate/9e6733ba-cad8-4624-b58f-416c93685e2c",
            "CertificateSource": "acm",
            "MinimumProtocolVersion": "TLSv1",
            "SSLSupportMethod": "sni-only"
          },
          "WebACLId": ""
        }
      }
    }
  },
  "changed": false,
  "msg": "Retrieved cloudfront facts."
}
```
